### PR TITLE
Add the grep command

### PR DESCRIPTION
**Pull request by nounoursheureux**
It only supports linear search for now.

### CODE REVIEW

**Comment by Ticki** [comment]
It doesn't compile on latest nightly. By the way, could you move this to redox-os/extrautils? I think it might fit better there.

File: src/bin/grep.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
+extern crate coreutils;
```
Comment (Ticki) [inline src/bin/grep.rs]: Could you add `#![deny(warnings)]` here?

File: src/bin/grep.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+fn main() {
+    let args = env::args().skip(1);
+    let stdout = io::stdout();
+    let mut stdout = stdout.lock();
+    let mut stderr = io::stderr();
+    let stdin = io::stdin();
+    let stdin = stdin.lock();
+
+    let mut flags = Flags::new();
+    let mut pattern = "".to_owned();
```
Comment (Ticki) [inline src/bin/grep.rs]: Use `String::with_capacity(N)` instead.

File: src/bin/grep.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    let mut files = vec![];
```
Comment (Ticki) [inline src/bin/grep.rs]: `Vec::new()` or `Vec::with_capacity(N)` here.

File: src/bin/grep.rs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                Err(e) => {
+                    stderr.writeln(e.description().as_bytes()).try(&mut stderr);
+                    return;
+                }
+            }
+        }
+    }
+
+    if &pattern == "" {
+        stderr.writeln("You must provide a pattern".as_bytes()).try(&mut stderr);
```
Comment (Ticki) [inline src/bin/grep.rs]: You can just prepend a `b` to the string and it will automatically be a byte string.

File: src/bin/grep.rs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                }
+            }
+        } else if &pattern == "" {
+            pattern = arg.clone();
+        } else {
+            match File::open(&Path::new(&arg)) {
+                Ok(f) => files.push(f),
+                Err(e) => {
+                    stderr.writeln(e.description().as_bytes()).try(&mut stderr);
+                    return;
```
Comment (Ticki) [inline src/bin/grep.rs]: This should exit with code, `1`.

File: src/bin/grep.rs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        if arg.starts_with("-") {
+            match arg.as_str() {
+                "-h" | "--help" => {
+                    stdout.writeln(MAN_PAGE.as_bytes()).try(&mut stderr);
+                },
+                "-n" | "--line-number" => flags.line_numbers = true,
+                _ => {
+                    stderr.write_all(b"Unkown option: ").try(&mut stderr);
+                    stderr.writeln(arg.as_bytes()).try(&mut stderr);
+                    return;
```
Comment (Ticki) [inline src/bin/grep.rs]: `exit(1)`

File: src/bin/grep.rs (hunk 5 above)
Comment (Ticki) [inline src/bin/grep.rs]: You can even choose to use `fail`.

File: src/bin/grep.rs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    if files.is_empty() {
+        do_linear_search(stdin, &pattern, &mut stdout, &mut stderr, flags);
+    } else {
+        for f in files {
+            do_linear_search(BufReader::new(f), &pattern, &mut stdout, &mut stderr, flags);
+        }
+    }
+}
+
+fn do_linear_search<T: BufRead, O: Write + WriteExt>(reader: T, pattern: &str, out: &mut O, stderr: &mut Stderr, flags: Flags) {
```
Comment (Ticki) [inline src/bin/grep.rs]: Actually, this is not linear search. So you might want to call this something else.

**Comment by Ticki** [comment]
Nice work, and well-written! :+1:

**Comment by Mouvedia** [comment]
Should we track the addition of the `-P` option in a new issue?

**Comment by Ticki** [comment]
@Mouvedia Yeah! You can open one in redox-os/extrautils.

**Pull request closed without merging**